apps/shell/tash_security.c

```diff
@@ -23,43 +23,167 @@
-#include <mbedtls/sha256.h>
+#include <mbedtls/pkcs5.h>
+#include <security/security_crypto.h>
 #include "tash_internal.h"
 
-#define SHA256_HASH_nCHAR  (64) // 32 bytes
-#define SHA256_HASH_nBYTE  (32) // 32 bytes
-#define CONVERT_nCHAR      (sizeof(unsigned long) * 2) // unsigned long for strtoul
+#define CONVERT_nCHAR			(sizeof(unsigned long) * 2) // unsigned long for strtoul
+
+#define PBKDF2_KEY_nCHAR		(64) // 32 bytes
+#define PBKDF2_KEY_LEN			(32) // 32 bytes
+#define PBKDF2_SALT_LEN			(16) // 16 bytes
+#define PBKDF2_SALT_nCHAR		(32) // 16 bytes
+#define AES_BLOCK_SIZE			(16)
+#define AES128_KEY			"ss/1"
+
+
+static int salt_aes_decrypt(char *input, unsigned int input_size, char *output, unsigned int output_size) 
+{
+	if ((input_size <= 0) || (input_size % AES_BLOCK_SIZE)) {
+		printf("Invalid AES key input size\n");
+		printf("input size : %d\n", input_size);
+		return -1;
+	}
+
+	if (input == NULL) {
+		printf("AES input is null\n");
+		return -1;
+	} 
+
+	if (output_size < input_size) {
+		printf("Invalid AES key output size\n");
+		printf("output size : %d\n", output_size);
+		return -1;
+	}
+
+	if (output == NULL) {
+		printf("AES output is null\n");
+		return -1;
+	}
+
+	security_handle hnd;
+	security_data encrypt_data;
+	security_data decrypt_data;
+	security_aes_param aes_param;
+	security_error ret;
+	char *tmp = NULL;
+
+	ret = security_init(&hnd);
+	if (ret != SECURITY_OK) {
+		printf("security_crypto : security_init failed. ret : %d\n", ret);
+		return -1;
+	}
+
+	encrypt_data.data = input;
+	encrypt_data.length = input_size;
+
+	decrypt_data.length = output_size;
+
+	aes_param.mode = AES_CBC_NOPAD;
+	aes_param.iv = NULL;
+	aes_param.iv_len = 0;
+
+	ret = crypto_aes_decryption(hnd, &aes_param, AES128_KEY, &encrypt_data, &decrypt_data);
+	if (ret != SECURITY_OK) {
+		printf("security_crypto : crypto_aes_decryption failed. ret : %d\n", ret);
+		return -1;
+	}
+
+	tmp = decrypt_data.data;
+	for (int i = 0; i < output_size; i++) {
+		output[i] = tmp[i];
+	}
+
+	return 1;
+}
+
+static int get_pbkdf2(char *input, int input_len, char *salt, int salt_len, int iteration, unsigned char *pbkdf2_output)
+{
+	mbedtls_md_context_t sha_ctx;
+	const mbedtls_md_info_t *info_sha;
+
+	/* Setup the hash/HMAC function, for the PBKDF2 function. */
+	mbedtls_md_init(&sha_ctx);
+
+	info_sha = mbedtls_md_info_from_type(MBEDTLS_MD_SHA256);
+	if (info_sha == NULL) {
+		printf("Failed to get hash information\n");
+		mbedtls_md_free(&sha_ctx);
+		return -1;
+	}
+
+	int ret = mbedtls_md_setup(&sha_ctx, info_sha, 1);
+	if (ret != 0) {
+		printf("Failed to setup hash function");
+		mbedtls_md_free(&sha_ctx);
+		return -1;
+	}
+
+	ret = mbedtls_pkcs5_pbkdf2_hmac(&sha_ctx, (const unsigned char *)input, input_len, (const unsigned char *)salt, salt_len, iteration, PBKDF2_KEY_LEN, pbkdf2_output);
+	if (ret != 0) {
+		printf("Call to mbedtls PBKDF2 function failed\n");
+		mbedtls_md_free(&sha_ctx);
+		return -1;
+	}
+
+	mbedtls_md_free(&sha_ctx);
+
```

Comment:
Would you consider using "goto" for all failure cases above to unify the mbedtls_md_free() calls?